[WIP] Data-structures tutorial (closes #57) (#59)
ProtoSchool/protoschool.github.io: src/lessons/Data-Structures/04.md

So far we've exclusively been discussing adorable images for the sake of our general
merriment, but content addressing can be used on all different types of files and data,
from JSON objects to term papers to videos. For cryptographic hashing to work, we need
to know what data format we're working with and use an appropriate tool.

## Decoding data structures

A `CID` is a particular form of content addressing. It's a specification developed for
IPFS, a specific decentralized web protocol that we'll discuss in later tutorials,
but it has very broad implications.

A `CID` is a single identifier that contains a cryptographic hash **and**
information about how to interpret that data called a "codec." (Codecs encode,

> decode, compress, and decompress data in certain formats. You use them every time

Author: I'd leave compress and decompress out for now. Technically, there's not compression in the current codecs, and we're trying to do compression at another layer (transport and storage). This will be quite confusing for anyone trying to map these semantics on to what they already know about HTTP so I'd just leave compression out for now.

> you stream a YouTube video of an adorable beagle.)

Author: Video codecs, like those on YouTube, are a different beast, so I'd leave this out.

In fact, we already have be aware of video codecs when encoding files in IPFS, just at a different layer (chunking).

```
+-------+------------------------------+
| Codec | Multihash                    |
+-------+------------------------------+
```

> Many formats and protocols use content addressing. `CID` allows us to create a universal identifier for any of these sytems. Tools like git and protocols like Etherium and Bitcoin all use content addressing already, but they differ greatly in how to interpret the data and what cryptographic function they use for hashing.

Author: ```suggestion
Many formats and protocols use content addressing. `CID` allows us to create a universal identifier for any of these systems. Tools like git and protocols like Ethereum and Bitcoin all use content addressing already, but they differ in how to interpret the data and what cryptographic function they use for hashing.
```

A `CID` is an identifier that contains the `codec` to interpret the data and a `multihash` which is a self-desribing hash (a hash that tells you what type of hashing function was used to create it).

```
+------------------------------+
|Codec                         |
+------------------------------+
|Multihash                     |
| +----------+---------------+ |
| |Hash Type | Hash Value    | |
| +----------+---------------+ |
|                              |
+------------------------------+
```

## Linking between different data structures

`CID`s allow us to build data structures that link to other data structures
in completely different formats. Imagine a tree of JSON objects that link
to BSON objects than also link to git commits. (Or imagine a file folder containing
puppy images <em>and</em> kitty videos, with a subfolder containing articles on
giraffes. The possiblities are endless!) All the way down this tree we
have cryptographic hashes that allow us to distribute and link the data.

> Why is important to link between different data structures? Every day on the centralized web,

Author: ```suggestion
Why is it important to link between different data structures? Every day on the centralized web,
```

we link from text to images, from logos to homepages, and from emails to PDFs.

> Links ties resources together, convey meaning, and make the web awesomely interactive!

Author: ```suggestion
Links tie resources together, convey meaning, and make the web awesomely interactive!
```
